editor/plugins/gradient_texture_2d_editor_plugin.cpp

```diff
@@ -0,0 +1,285 @@
+
+#include "editor/editor_node.h"
+#include "editor/editor_scale.h"
+#include "scene/gui/box_container.h"
+#include "scene/gui/flow_container.h"
+#include "scene/gui/separator.h"
+
+Point2 GradientTexture2DEditorRect::_get_handle_position(const Handle p_handle) {
+	// Get the handle's mouse position in pixels relative to offset.
+	const Vector2 percent = p_handle == HANDLE_FILL_FROM ? texture->get_fill_from() : texture->get_fill_to();
+	return Vector2(CLAMP(percent.x, 0, 1), CLAMP(percent.y, 0, 1)) * size;
+}
+
+void GradientTexture2DEditorRect::_update_fill_position() {
+	if (handle == HANDLE_NONE) {
+		return;
+	}
+
+	// Update the texture's fill_from/fill_to property based on mouse input.
+	Vector2 percent = (get_local_mouse_position() - offset) / size;
+	percent = Vector2(CLAMP(percent.x, 0, 1), CLAMP(percent.y, 0, 1));
+	if (snap_enabled) {
+		percent = Vector2(Math::snapped(percent.x - 0.5, snap_size) + 0.5, Math::snapped(percent.y - 0.5, snap_size) + 0.5);
+	}
+
+	String property_name = handle == HANDLE_FILL_FROM ? "fill_from" : "fill_to";
+
+	undo_redo->create_action(vformat(TTR("Set %s"), property_name), UndoRedo::MERGE_ENDS);
+	undo_redo->add_do_property(texture.ptr(), property_name, percent);
+	undo_redo->add_undo_property(texture.ptr(), property_name, handle == HANDLE_FILL_FROM ? texture->get_fill_from() : texture->get_fill_to());
+	undo_redo->commit_action();
+}
+
+void GradientTexture2DEditorRect::gui_input(const Ref<InputEvent> &p_event) {
+	// Grab/release handle.
+	const Ref<InputEventMouseButton> mb = p_event;
+	if (mb.is_valid() && mb->get_button_index() == MouseButton::LEFT) {
+		if (mb->is_pressed()) {
+			Point2 mouse_position = mb->get_position() - offset;
+			if (Rect2(_get_handle_position(HANDLE_FILL_FROM).round() - handle_size / 2, handle_size).has_point(mouse_position)) {
+				handle = HANDLE_FILL_FROM;
+			} else if (Rect2(_get_handle_position(HANDLE_FILL_TO).round() - handle_size / 2, handle_size).has_point(mouse_position)) {
+				handle = HANDLE_FILL_TO;
+			} else {
+				handle = HANDLE_NONE;
+			}
+		} else {
+			_update_fill_position();
+			handle = HANDLE_NONE;
+		}
+	}
+
+	// Move handle.
+	const Ref<InputEventMouseMotion> mm = p_event;
+	if (mm.is_valid()) {
+		_update_fill_position();
+	}
+}
+
+void GradientTexture2DEditorRect::set_texture(Ref<GradientTexture2D> &p_texture) {
+	texture = p_texture;
+	texture->connect("changed", callable_mp((CanvasItem *)this, &CanvasItem::update));
+}
+
+void GradientTexture2DEditorRect::set_snap_enabled(bool p_snap_enabled) {
+	snap_enabled = p_snap_enabled;
+	update();
+}
+
+void GradientTexture2DEditorRect::set_snap_size(float p_snap_size) {
+	snap_size = p_snap_size;
+	update();
+}
+
+void GradientTexture2DEditorRect::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE:
+		case NOTIFICATION_THEME_CHANGED: {
+			checkerboard->set_texture(get_theme_icon(SNAME("GuiMiniCheckerboard"), SNAME("EditorIcons")));
+		} break;
```

Comment:
```suggestion
		} break;

```

Separation lind between breaks in big `_notification` switches for readability.